scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java

```diff
@@ -12,125 +13,477 @@
+         *             only be changed to {@code BLOB} data type.
+         *             In this case, the CDC log schema will have
+         *             this column with {@code BLOB} and this
+         *             method will not get called.</td>
+         *     </tr>
+         * </table>
+         *
+         * @param columnName the column name to check if it is a <code>LIST</code>
+         * @return whether the column is a <code>LIST</code> in the base table.
+         * @throws UnresolvableSchemaInconsistencyException if whether the column
+         *                                                is a <code>LIST</code> cannot
+         *                                                be deduced.
+         */
+        public boolean isList(String columnName) throws UnresolvableSchemaInconsistencyException {
+            ColumnMetadata baseColumnMetadata = baseTableSchema.getColumn(columnName);
+            if (baseColumnMetadata == null) {
+                throw new UnresolvableSchemaInconsistencyException(
+                        String.format("Column '%s' in base table metadata is missing, " +
+                                "making it impossible to deduce whether it was a LIST at " +
+                                "a time of CDC log schema.", columnName));
+            }
+
+            boolean isList = baseColumnMetadata.getType().getName() == DataType.Name.LIST;
+            boolean isMapTimeuuid = baseColumnMetadata.getType().getName() == DataType.Name.MAP
+                    && baseColumnMetadata.getType().getTypeArguments().get(0).getName() == DataType.Name.TIMEUUID;
+
+            if (!isList && !isMapTimeuuid) {
+                throw new UnresolvableSchemaInconsistencyException(
+                        String.format("The type of column '%s' in base table metadata is %s, " +
+                                "making it impossible to deduce whether it was a LIST at " +
+                                "a time of CDC log schema.", columnName, baseColumnMetadata.getType().getName()));
+            }
+
+            return isList;
+        }
     }
 
-    private ChangeSchema.DataType translateColumnDataType(DataType driverType) {
+    /**
+     * Thrown when the {@link Driver3SchemaBuilder} cannot resolve
+     * a schema inconsistency between the base table schema
+     * and CDC log schema.
+     */
+    public static class UnresolvableSchemaInconsistencyException extends Exception {
+        private UnresolvableSchemaInconsistencyException(String message) {
+            super(message);
+        }
+    }
+
+    private ChangeSchema.DataType translateCdcColumnDataType(DataType driverType) {
+        // This method only translates driverType
+        // from CDC table columns. All CDC table columns (with non-native types)
+        // are frozen, even though the driverType can lack this information
+        // (when driverType is from a PreparedStatement metadata).
```

Comment:
Could you explain a bit more what's happening with the prepared statements that's special?

Reply:
In this specific line: We get the CDC log schema from `cdcRow.getColumnDefinitions()`. This object does not have information about whether the types are frozen or not (so they look like non-frozen). However, if you used the metadata from `cluster.getMetadata()`, it'd contain the information about whether the types are frozen or not.